sdk/storage/Azure.Storage.Common/tests/Shared/PredictableStream.cs

```diff
@@ -18,14 +18,18 @@ internal class PredictableStream : Stream
 
         public override bool CanWrite => false;
 
+        private readonly bool _revealsLength;
         private readonly long? _length;
-        public override long Length => _length ?? throw new NotSupportedException();
+        public override long Length => _revealsLength
+            ? _length ?? throw new NotSupportedException()
+            : throw new NotSupportedException();
```

Comment:
So in prod code we decided to use inheritance to model seekability vs non-seekability (SeekableWindowStream and UnseekableWindowStream). But in tests we drive this through constructor flag (i.e. we don't have SeekablePredicatableStream). 

I was about to make a comment that we could just accept flag `seekable` in WindowStream instead of creating so many types, but I thought it's gonna be lot of if/else branching logic. But then I see this code and I think it's more concise than inheritance based solution. If you don't mind I'd try to make WindowStream accept a flag and see if it would look better (less complex).

Reply:
I chose to split the implementations because there were differences in managing state instead of just when to pass things through. `UnseekableWindowStream` has to maintain bytes read and an overall max possible bytes before finishing the stream facade. These are useless in `SeekableWindowStream`, which operates on public `Length` properties and has no use for a position. Perhaps we want to try a composition approach, where `WindowStream` implements what it currently does and contains some implementation of the rest of these methods? I still think the if/else across all these methods would be a bit much to read, especially when they're all branching off the same condition.

Reply:
Yeah, I can imagine this being too much of spaghetti with these ifs. Your call.